aten/src/ATen/native/cpu/batch_norm_kernel.cpp

```diff
@@ -821,74 +915,93 @@ inline void batch_norm_cpu_collect_stats_channels_last_internal(
+      TORCH_CHECK(tid < num_threads, "expect thread id smaller than ", num_threads, ", got thread id ", tid);
+      opmath_t* buffer_ptr = buffer_data + tid * n_channel;
+      for (const auto i : c10::irange(begin, end)) {
+        const scalar_t* input_ptr = input_data + i * n_channel;
+        int64_t d = 0;
+        for (; d < n_channel - (n_channel % bVec::size()); d += bVec::size()) {
+          bVec data_bvec = bVec::loadu(input_ptr + d);
+          fVec data_fvec0, data_fvec1;
+          std::tie(data_fvec0, data_fvec1) = convert_to_float<scalar_t>(data_bvec);
+          fVec mean_fvec0, mean_fvec1;
+          std::tie(mean_fvec0, mean_fvec1) = load2f(mean_data + d);
+          fVec var_fvec0 = fVec::loadu(buffer_ptr + d);
+          fVec var_fvec1 = fVec::loadu(buffer_ptr + d + fVec::size());
+          var_fvec0 += (data_fvec0 - mean_fvec0) * (data_fvec0 - mean_fvec0);
+          var_fvec1 += (data_fvec1 - mean_fvec1) * (data_fvec1 - mean_fvec1);
+          var_fvec0.store(buffer_ptr + d);
+          var_fvec1.store(buffer_ptr + d + fVec::size());
+        }
+        for (; d < n_channel; d++) {
+          opmath_t data_val = opmath_t(input_ptr[d]);
+          opmath_t mean_val = opmath_t(mean_data[d]);
+          buffer_ptr[d] += (data_val - mean_val) * (data_val - mean_val);
+        }
+      }
+    });
 
-  buffer.zero_();
-  at::parallel_for(0, N, 1, [&](int64_t begin, int64_t end) {
-    int tid = at::get_thread_num();
-    TORCH_CHECK(tid < num_threads, "expect thread id smaller than ", num_threads, ", got thread id ", tid);
-    opmath_t* buffer_ptr = buffer_data + tid * n_channel;
-    for (const auto i : c10::irange(begin, end)) {
-      const scalar_t* input_ptr = input_data + i * n_channel;
-      int64_t d = 0;
-      for (; d < n_channel - (n_channel % bVec::size()); d += bVec::size()) {
-        bVec data_bvec = bVec::loadu(input_ptr + d);
-        fVec data_fvec0, data_fvec1;
-        std::tie(data_fvec0, data_fvec1) = convert_to_float<scalar_t>(data_bvec);
-        fVec mean_fvec0, mean_fvec1;
-        std::tie(mean_fvec0, mean_fvec1) = load2f(mean_data + d);
-        fVec var_fvec0 = fVec::loadu(buffer_ptr + d);
-        fVec var_fvec1 = fVec::loadu(buffer_ptr + d + fVec::size());
-        var_fvec0 += (data_fvec0 - mean_fvec0) * (data_fvec0 - mean_fvec0);
-        var_fvec1 += (data_fvec1 - mean_fvec1) * (data_fvec1 - mean_fvec1);
-        var_fvec0.store(buffer_ptr + d);
-        var_fvec1.store(buffer_ptr + d + fVec::size());
+    for (const auto c : c10::irange(n_channel)) {
+      opmath_t _var_sum = 0;
+      for (const auto t : c10::irange(num_threads)) {
+        _var_sum += buffer_data[t * n_channel + c];
       }
-      for (; d < n_channel; d++) {
-        opmath_t data_val = opmath_t(input_ptr[d]);
-        opmath_t mean_val = opmath_t(mean_data[d]);
-        buffer_ptr[d] += (data_val - mean_val) * (data_val - mean_val);
+      var_sum_data[c] = param_t(_var_sum);
+    }
+  } else {
```

Comment:
@jgong5, should I separate the PR for `batch_norm_cpu_collect_stats_channels_last_internal`?

Reply:
That's fine. Up to you.

Reply:
I will apply the same heuristics and re-measure perf for `batch_norm_cpu_collect_stats_channels_last_internal` in a separate PR.